File: 算法/代码随想录算法PDF/00调试源代码/459.重复的子字符串.cpp
/*
 * @lc app=leetcode.cn id=459 lang=cpp
 *
 * [459] 重复的子字符串
 */

// @lc code=start
class Solution {
public:
    void getNext(int* next, string s) {
        int j = -1;
        next[0] = j;
        for (int i = 1; i < s.size(); i++) { // 遍历s填充next
            while (j >= 0 && s[j + 1] != s[i]) {
                j = next[j];  // 回溯相同子串继续匹配
            }
            if (s[j + 1] == s[i]) {
                j++;
            }
            next[i] = j;
        }
    }
    bool repeatedSubstringPattern(string s) {
        if (s.size() == 0) {  // 特殊情况
            return false;
        }
        int next[s.size()];
        getNext(next, s);
        if (next[s.size() - 1] != -1 && s.size() % (s.size() - (next[s.size() - 1] + 1)) == 0) {
            return true;
        } else {
            return false;
        }
    }
};
// @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/459.重复的子字符串.py
#
# @lc app=leetcode.cn id=459 lang=python3
#
# [459] 重复的子字符串
#

# @lc code=start
class Solution:
    def repeatedSubstringPattern(self, s: str) -> bool:
        def getNext(s):
            next = [0] * len(s)
            j = -1
            next[0] = j
            for i in range(1, len(s)):
                while j >= 0 and s[i] != s[j + 1]:
                    j = next[j]
                if s[i] == s[j + 1]:
                    j += 1
                next[i] = j
            return next
        if len(s) == 0:
            return False
        next = getNext(s)
        if next[-1] != -1 and len(s) % (len(s) - (next[-1] + 1)) == 0:
            return True
        return False
# @lc code=end

